File: org.fantom.fwt/source/js/SliderPeer.js
//
// Copyright (c) 2011, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   31 Jan 11  Ivan Kornienko  Creation
//

fan.fwt.SliderPeer = fan.sys.Obj.$extend(fan.fwt.WidgetPeer);
fan.fwt.SliderPeer.prototype.$ctor = function(self) {}

fan.fwt.SliderPeer.prototype.val   = function(self) { return this.m_val; }
fan.fwt.SliderPeer.prototype.val$  = function(self, val)
{
  this.m_val = val;
  if (this.elem != null) this.sync(self);
} 
fan.fwt.SliderPeer.prototype.m_val = 0;

fan.fwt.SliderPeer.prototype.min   = function(self) { return this.m_min; }
fan.fwt.SliderPeer.prototype.min$  = function(self, val)
{ 
  // for compatibility with Java SWT peer
  if (val < 0 || val >= this.m_max) return;
  this.m_min = val;
  if (this.elem != null) this.sync(self);
}
fan.fwt.SliderPeer.prototype.m_min = 0;

fan.fwt.SliderPeer.prototype.max   = function(self) { return this.m_max; }
fan.fwt.SliderPeer.prototype.max$  = function(self, val) 
{ 
  // for compatibility with Java SWT peer
  if (val < 0 || val <= this.m_min) return;
  this.m_max = val;
  if (this.elem != null) this.sync(self);
}

fan.fwt.SliderPeer.prototype.m_max = 100;

fan.fwt.SliderPeer.prototype.thumb  = function(self) { return this.m_thumb; }
fan.fwt.SliderPeer.prototype.thumb$ = function(self, val)
{
  this.m_thumb = val;
  if (this.elem != null) this.sync(self);
}
fan.fwt.SliderPeer.prototype.m_thumb = 10;

// really page option isn't supported
fan.fwt.SliderPeer.prototype.page   = function(self) { return this.m_page; }
fan.fwt.SliderPeer.prototype.page$  = function(self, val) { this.m_page = val; } 
fan.fwt.SliderPeer.prototype.m_page = 10;

fan.fwt.SliderPeer.prototype.prefSize = function(self, hints)
{
  var pref = fan.fwt.WidgetPeer.prototype.prefSize.call(this, self, hints);
  var thickness = fan.fwt.SliderPeer.thickness();
  if (self.m_orientation == fan.gfx.Orientation.m_horizontal)
    return fan.gfx.Size.make(pref.m_w, thickness);
  else
    return fan.gfx.Size.make(thickness, pref.m_h);
}

fan.fwt.SliderPeer.prototype.create = function(parentElem, self)
{
  var scrollDiv = document.createElement("div");
  scrollDiv.style.padding = "0px";       
  var scrollContent = document.createElement("div");
  scrollDiv.appendChild(scrollContent);

  var vertical = self.m_orientation == fan.gfx.Orientation.m_vertical;
  if (vertical)
  {
    scrollDiv.style.width = fan.fwt.SliderPeer.thickness() + "px";
    scrollDiv.style.overflowX = "hidden";
    scrollDiv.style.overflowY = "scroll";
    scrollContent.style.width = "1px";
  }
  else
  {
    scrollDiv.style.height = fan.fwt.SliderPeer.thickness() + "px";
    scrollDiv.style.overflowX = "scroll";
    scrollDiv.style.overflowY = "hidden";
    scrollContent.style.height = "1px";
  }

  scrollDiv.onscroll = function(event)
  { 
    var scrollIndent = 0;
    var scrollSize = 0;
    if (self.m_orientation == fan.gfx.Orientation.m_horizontal)
    {
      scrollSize = scrollDiv.scrollWidth - scrollDiv.clientWidth;
      scrollIndent = scrollDiv.scrollLeft;
    }  
    else
    {
      scrollSize = scrollDiv.scrollHeight - scrollDiv.clientHeight;
      scrollIndent = scrollDiv.scrollTop;
    }
    var newVal = Math.round(self.peer.m_min + scrollIndent * (self.peer.m_max - self.peer.m_min - self.peer.m_thumb) / scrollSize);
    if (self.peer.m_val == newVal)
      return

    self.peer.m_val = newVal;
    // fire onModify
    if (self.m_onModify.size() > 0)
    {
      var me = fan.fwt.Event.make();
      me.m_id = fan.fwt.EventId.m_modified;
      me.m_widget = self;
      var list = self.m_onModify.list();
      for (var i=0; i<list.size(); i++) list.get(i).call(me);
    }
  }

  // container element
  var div = this.emptyDiv();
  div.appendChild(scrollDiv);
  parentElem.appendChild(div);
  return div;
}

fan.fwt.SliderPeer.prototype.sync = function(self)
{
  var vert = self.m_orientation == fan.gfx.Orientation.m_vertical;
  var w = this.m_size.m_w;
  var h = this.m_size.m_h;
  var scrollDiv = this.elem.firstChild;
  var scrollContent = scrollDiv.firstChild;

  var maxRatio = (this.m_max - this.m_min) / this.m_thumb;
  var valRatio = (this.m_val - this.m_min) / this.m_thumb;

  if (vert)
  {
    scrollDiv.style.height = h + "px";
    if (this.m_enabled)
    {
      scrollContent.style.height = Math.round(h * maxRatio) + "px";
      scrollDiv.scrollTop = Math.round(h * valRatio);
    }
    else
    {
      scrollDiv.scrollTop = 0;
      scrollContent.style.height = "0px";
    }
  }
  else
  {
    scrollDiv.style.width = w + "px";
    if (this.m_enabled)
    {
      scrollContent.style.width = Math.round(w * maxRatio) + "px";
      scrollDiv.scrollLeft = Math.round(w * valRatio);
    }
    else
    {
      scrollDiv.scrollLeft = 0;
      scrollContent.style.width = "0px";
    }
  }

  fan.fwt.WidgetPeer.prototype.sync.call(this, self);
}

fan.fwt.SliderPeer.prototype.checkModifyListeners = function(self) {}

//////////////////////////////////////////////////////////////////////////
// Utils
//////////////////////////////////////////////////////////////////////////

fan.fwt.SliderPeer.thickness = function()
{
  if (fan.fwt.SliderPeer.m_thickness == 0)
  {
    var inner = document.createElement('div');
    inner.style.height = "100px";

    var outer = document.createElement('div');
    with (outer.style)
    {
      width = "50px"; height = "50px";
      overflow = "hidden"; position = "absolute";
      visibility = "hidden";
    }
    outer.appendChild(inner);

    document.body.appendChild(outer);
    var w1 = inner.offsetWidth;
    outer.style.overflow = 'scroll';
    var w2 = inner.offsetWidth;
    if (w1 == w2) w2 = outer.clientWidth;
    document.body.removeChild(outer);

    fan.fwt.SliderPeer.m_thickness = (w1 - w2);
  }
  return fan.fwt.SliderPeer.m_thickness;
}

fan.fwt.SliderPeer.m_thickness = 0;

File: org.fantom.fwt/source/java/SliderPeer.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   16 Nov 10  Yuri Strot  Creation
//
package fan.fwt;

import fan.sys.*;
import fan.gfx.Orientation;
import org.eclipse.swt.*;
import org.eclipse.swt.widgets.*;
import org.eclipse.swt.widgets.Slider;
import org.eclipse.swt.widgets.Widget;
import org.eclipse.swt.events.*;

public class SliderPeer
  extends WidgetPeer
  implements SelectionListener
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static SliderPeer make(fan.fwt.Slider self)
    throws Exception
  {
    SliderPeer peer = new SliderPeer();
    ((fan.fwt.Widget)self).peer = peer;
    peer.self = self;
    return peer;
  }

  public Widget create(Widget parent)
  {
    fan.fwt.Slider self = (fan.fwt.Slider)this.self;
	int style = self.orientation == Orientation.horizontal ? SWT.HORIZONTAL : SWT.VERTICAL;
    Slider c = new Slider((Composite)parent, style);
    c.addSelectionListener(this);
    this.control = c;
    return c;
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  // Int val := 0
  public long val(fan.fwt.Slider self) { return val.get(); }
  public void val(fan.fwt.Slider self, long v) { val.set(v); }
  public final Prop.IntProp val = new Prop.IntProp(this, 0)
  {
    public int get(Widget w) { return ((Slider)w).getSelection(); }
    public void set(Widget w, int v) { ((Slider)w).setSelection(v); }
  };

  // Int min := 0
  public long min(fan.fwt.Slider self) { return min.get(); }
  public void min(fan.fwt.Slider self, long v) { min.set(v); }
  public final Prop.IntProp min = new Prop.IntProp(this, 0)
  {
    public int get(Widget w) { return ((Slider)w).getMinimum(); }
    public void set(Widget w, int v) { ((Slider)w).setMinimum(v); }
  };

  // Int max := 100
  public long max(fan.fwt.Slider self) { return max.get(); }
  public void max(fan.fwt.Slider self, long v) { max.set(v); }
  public final Prop.IntProp max = new Prop.IntProp(this, 100)
  {
    public int get(Widget w) { return ((Slider)w).getMaximum(); }
    public void set(Widget w, int v) { ((Slider)w).setMaximum(v); }
  };

  // Int thumb := 10
  public long thumb(fan.fwt.Slider self) { return thumb.get(); }
  public void thumb(fan.fwt.Slider self, long v) { thumb.set(v); }
  public final Prop.IntProp thumb = new Prop.IntProp(this, 10)
  {
    public int get(Widget w) { return ((Slider)w).getThumb(); }
    public void set(Widget w, int v) { ((Slider)w).setThumb(v); }
  };

  // Int page := 10
  public long page(fan.fwt.Slider self) { return page.get(); }
  public void page(fan.fwt.Slider self, long v) { page.set(v); }
  public final Prop.IntProp page = new Prop.IntProp(this, 10)
  {
    public int get(Widget w) { return ((Slider)w).getPageIncrement(); }
    public void set(Widget w, int v) { ((Slider)w).setPageIncrement(v); }
  };

//////////////////////////////////////////////////////////////////////////
// Eventing
//////////////////////////////////////////////////////////////////////////

  public void widgetDefaultSelected(SelectionEvent se) {}

  public void widgetSelected(SelectionEvent se)
  {
    fireModified();
  }

  public void fireModified()
  {
    Slider sb = (Slider)control;
    fan.fwt.Event fe = event(EventId.modified);
    fe.data = Long.valueOf(sb.getSelection());
    ((fan.fwt.Slider)self).onModify().fire(fe);
  }

  public void checkModifyListeners(fan.fwt.Slider self)
  {
  }

}